### DISCUSSION TITLE
Does it possible to update :devtools option using build hook?

### DISCUSSION THREAD

**Discussion by maksimr** [thheller/shadow-cljs#987] [Q&A] [answered]
I want to dynamically generate value for `:devtools-url` setting from `shadow-cljs.edn`. And trying to use hook instead of passing it as `--config-merge` option or global env variable, but looks like it doesn't work. 

Thnx

The build hook code:
```clj
(ns script.build
  (:require
   [clojure.string :refer [trim]]
   [clojure.java.shell :as shell]))

(defn setup
  "Gitpod generate URL for open port so we should
   use this URL to properly forward connection from browser to the server"
  {:shadow.build/stage :configure}
  [build-state]
  (if (not-empty (System/getenv "GITPOD_WORKSPACE_URL"))
    (assoc-in build-state [:devtools :devtools-url]
              (trim (:out (shell/sh "gp" "url" "9630"))))
    build-state))
```

**Comment by thheller** [comment, marked as answer]
At the `:configure` stage the `:devtools-url` will already be applied to the place it is actually used. So updating the place where it came from will do nothing. Basically all this sets is this `:closure-defines` key/val pair.

https://github.com/thheller/shadow-cljs/blob/c28859a8034bdeec3bd5c40846872c22fbcdd0f4/src/main/shadow/build/targets/shared.clj#L219-L220

So in your build hook you can `(assoc-in build-state [:compiler-options :closure-defines 'shadow.cljs.devtools.client.env/devtools-url] new-url)`.
